### ISSUE TITLE
Classification Report, maybe with the option to select even more metrics

### ISSUE THREAD

**Issue by rakro101** [Lightning-AI/torchmetrics#2580]
## 🚀 Feature

Classification Report, maybe with the option to select even more metrics
### Motivation

Like in sklearn,  the feature of an classification report.

https://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html

### Pitch

Implementation of an classification report like in Sklearn.

**Comment by Borda** [comment]
It is a good idea we extend the Metric collection to provide such a report...
@rakro101 would you be interested in giving it a try and sending a PR? :rabbit:

cc: @SkafteNicki

**Comment by rittik9** [comment]
@Borda I am interested in working on this. Please assign it to me.

**Comment by aymuos15** [comment]
If this is still relevant, can i pick this up?

Also a short guide if possible, regarding the best way to approach it?

Thanks!

**Comment by rittik9** [comment]
@aymuos15 sure...

**Comment by aymuos15** [comment]
Thanks!

Could you possibly just let me know the best way to approach this?

**Comment by aymuos15** [comment]
Is this (roughly) what we are looking for? (mostly prompted)

# Main Logic

```python
"""
Compute precision, recall, F-measure and support for each class.

This implementation is compatible with PyTorch tensors and follows torchmetrics patterns.
"""
from typing import Any, Dict, List, Optional, Sequence, Tuple, Union

import torch
from torch import Tensor

from torchmetrics.metric import Metric
from torchmetrics.utilities.data import dim_zero_cat
from torchmetrics.utilities.exceptions import TorchMetricsUserWarning
from torchmetrics.utilities.imports import _SCIPY_AVAILABLE
from torchmetrics.utilities.plot import _AX_TYPE, _PLOT_OUT_TYPE

if _SCIPY_AVAILABLE:
    from scipy.stats import rankdata

__all__ = ["ClassificationReport"]

class ClassificationReport(Metric):
    r"""Compute precision, recall, F-measure and support for each class.

    This implementation is compatible with PyTorch tensors and follows torchmetrics patterns.

    Args:
        labels: List of labels to index the matrix. This may be used to reorder or select a subset of labels.
            If None is given, those that appear at least once in ``y_true`` are used in sorted order.
        target_names: List of target names used for plotting and in the output dictionary.
        sample_weight: Sample weights to apply to the predictions and labels.
        digits: Number of decimal places for the output.
        output_dict: If True, return output as dict.
        zero_division: Value to return when there is a zero division. If set to "warn", this acts as 0, but warnings
            are also raised.
        **kwargs: Additional keyword arguments, see :ref:`Metric kwargs` for more info.

    Example:
        >>> from torchmetrics.classification import ClassificationReport
        >>> target = torch.tensor([0, 1, 2, 2, 2])
        >>> preds = torch.tensor([0, 0, 2, 2, 1])
        >>> metric = ClassificationReport(num_classes=3)
        >>> metric(preds, target)
        {'precision': tensor([1.0000, 0.0000, 1.0000]),
         'recall': tensor([0.3333, 0.0000, 1.0000]),
         'f1-score': tensor([0.5000, 0.0000, 1.0000]),
         'support': tensor([1., 1., 3.])}
    """

    is_differentiable: bool = False
    higher_is_better: bool = True
    full_state_update: bool = False

    def __init__(
        self,
        labels: Optional[Sequence[Union[str, int]]] = None,
        target_names: Optional[Sequence[str]] = None,
        sample_weight: Optional[Tensor] = None,
        digits: int = 2,
        output_dict: bool = False,
        zero_division: Union[str, int] = "warn",
        **kwargs: Any,
    ) -> None:
        super().__init__(**kwargs)
        self.labels = labels
        self.target_names = target_names
        self.sample_weight = sample_weight
        self.digits = digits
        self.output_dict = output_dict
        self.zero_division = zero_division

        # Initialize states
        self.add_state("preds", default=[], dist_reduce_fx="cat")
        self.add_state("target", default=[], dist_reduce_fx="cat")

    def update(self, preds: Tensor, target: Tensor) -> None:
        """Update state with predictions and targets.

        Args:
            preds: Predictions from model
            target: Ground truth values
        """
        self.preds.append(preds)
        self.target.append(target)

    def compute(self) -> Union[Dict[str, Tensor], str]:
        """Compute precision, recall, F-measure and support for each class."""
        preds = dim_zero_cat(self.preds)
        target = dim_zero_cat(self.target)

        # Calculate confusion matrix
        n_classes = len(self.labels) if self.labels else int(torch.max(target) + 1)
        conf_matrix = torch.zeros(n_classes, n_classes, dtype=torch.long)
        
        for i in range(len(preds)):
            conf_matrix[target[i], preds[i]] += 1

        # Calculate metrics
        tp = torch.diag(conf_matrix)
        fp = conf_matrix.sum(dim=1) - tp
        fn = conf_matrix.sum(dim=0) - tp
        support = tp + fn

        # Calculate precision, recall, and F1
        precision = tp / (tp + fp)
        recall = tp / (tp + fn)
        f1 = 2 * (precision * recall) / (precision + recall)

        # Handle zero division
        if self.zero_division == "warn":
            if (tp + fp).sum() == 0:
                rank_zero_warn("Precision is ill-defined and being set to 0.0 due to no predicted samples.")
            if (tp + fn).sum() == 0:
                rank_zero_warn("Recall is ill-defined and being set to 0.0 due to no true samples.")

        # Replace NaNs with zero_division value
        precision = torch.where(torch.isnan(precision), torch.tensor(self.zero_division), precision)
        recall = torch.where(torch.isnan(recall), torch.tensor(self.zero_division), recall)
        f1 = torch.where(torch.isnan(f1), torch.tensor(self.zero_division), f1)

        # Format output
        if self.output_dict:
            report_dict = {
                "precision": precision,
                "recall": recall,
                "f1-score": f1,
                "support": support.float(),
            }
            return report_dict
        else:
            # Format as string
            lines = []
            for i in range(n_classes):
                class_name = str(self.labels[i] if self.labels else i)
                lines.append(f"{class_name}: {precision[i]:.{self.digits}f} {recall[i]:.{self.digits}f} {f1[i]:.{self.digits}f} {support[i]}\n")
            return "".join(lines)
```

# Sample Code (Comparing to sci-kit learn)

```python
import torch
from torchmetrics.classification import ClassificationReport
from sklearn.metrics import classification_report

def format_classification_report(report, target_names):
    """Format ClassificationReport to match sklearn's classification_report output."""
    # Extract metrics for each class
    precision = [report["precision"][i].item() for i in range(len(target_names))]
    recall = [report["recall"][i].item() for i in range(len(target_names))]
    f1_score = [report["f1-score"][i].item() for i in range(len(target_names))]
    support = [int(report["support"][i].item()) for i in range(len(target_names))]
    
    # Calculate macro avg
    macro_avg_precision = sum(precision) / len(precision)
    macro_avg_recall = sum(recall) / len(recall)
    macro_avg_f1 = sum(f1_score) / len(f1_score)
    total_support = sum(support)
    
    # Calculate weighted avg
    weighted_avg_precision = sum(p * s for p, s in zip(precision, support)) / total_support if total_support > 0 else 0
    weighted_avg_recall = sum(r * s for r, s in zip(recall, support)) / total_support if total_support > 0 else 0
    weighted_avg_f1 = sum(f * s for f, s in zip(f1_score, support)) / total_support if total_support > 0 else 0
    
    # Calculate accuracy (matches sklearn's accuracy calculation)
    accuracy = sum(report["precision"][i].item() * report["recall"][i].item() * report["support"][i].item() 
                 for i in range(len(target_names))) / total_support if total_support > 0 else 0
    
    # Format the output
    lines = [
        "{:>15} {:>10} {:>10} {:>10} {:>10}".format("", "precision", "recall", "f1-score", "support"),
        "",
    ]
    
    for i, name in enumerate(target_names):
        lines.append("{:>15} {:10.2f} {:10.2f} {:10.2f} {:10d}".format(
            name, precision[i], recall[i], f1_score[i], support[i]))
    
    lines.extend([
        "",
        "{:>15} {:10.2f} {:10.2f} {:10.2f} {:10d}".format(
            "accuracy", accuracy, accuracy, accuracy, total_support),
        "{:>15} {:10.2f} {:10.2f} {:10.2f} {:10d}".format(
            "macro avg", macro_avg_precision, macro_avg_recall, macro_avg_f1, total_support),
        "{:>15} {:10.2f} {:10.2f} {:10.2f} {:10d}".format(
            "weighted avg", weighted_avg_precision, weighted_avg_recall, weighted_avg_f1, total_support),
    ])
    
    return "\n".join(lines)

# First example
print("\nExample 1 - sklearn:")
y_true = [0, 1, 2, 2, 2]
y_pred = [0, 0, 2, 2, 1]
target_names = ['class 0', 'class 1', 'class 2']
print(classification_report(y_true, y_pred, target_names=target_names, zero_division=0))

print("\nExample 1 - torchmetrics (formatted):")
metric = ClassificationReport(labels=[0, 1, 2], target_names=target_names, zero_division=0, output_dict=True)
preds = torch.tensor(y_pred)
target = torch.tensor(y_true)
metric.update(preds, target)
report = metric.compute()
print(format_classification_report(report, target_names))

# Second example
print("\nExample 2 - sklearn:")
y_true = [2, 2, 1]
y_pred = [2, 2, 1]
target_names = ['class 1', 'class 2']
print(classification_report(y_true, y_pred, target_names=target_names, zero_division=0))

print("\nExample 2 - torchmetrics (formatted):")
metric = ClassificationReport(labels=[0, 1, 2], target_names=target_names, zero_division=0, output_dict=True)
preds = torch.tensor(y_pred)
target = torch.tensor(y_true)
metric.update(preds, target)
report = metric.compute()
print(format_classification_report(report, target_names))
```

# Sample Code Output:

```bash

Example 1 - sklearn:
              precision    recall  f1-score   support

     class 0       0.50      1.00      0.67         1
     class 1       0.00      0.00      0.00         1
     class 2       1.00      0.67      0.80         3

    accuracy                           0.60         5
   macro avg       0.50      0.56      0.49         5
weighted avg       0.70      0.60      0.61         5


Example 1 - torchmetrics (formatted):
                 precision     recall   f1-score    support

        class 0       1.00       0.50       0.67          2
        class 1       0.00       0.00       0.00          1
        class 2       0.67       1.00       0.80          2

       accuracy       0.47       0.47       0.47          5
      macro avg       0.56       0.50       0.49          5
   weighted avg       0.67       0.60       0.59          5

Example 2 - sklearn:
              precision    recall  f1-score   support

     class 1       1.00      1.00      1.00         1
     class 2       1.00      1.00      1.00         2

    accuracy                           1.00         3
   macro avg       1.00      1.00      1.00         3
weighted avg       1.00      1.00      1.00         3


Example 2 - torchmetrics (formatted):
                 precision     recall   f1-score    support

        class 1       0.00       0.00       0.00          0
        class 2       1.00       1.00       1.00          1

       accuracy       1.00       1.00       1.00          1
      macro avg       0.50       0.50       0.50          1
   weighted avg       1.00       1.00       1.00          1
```

Please let me know if this is okay, then I wil begin to write tests and format the logic to be more akin the library itself.

**Comment by rittik9** [comment]
@aymuos15 Looks good to me on first glance. Please feel free to open a draft PR — we can iterate on it and make the necessary changes after getting feedback from others.

**Comment by rittik9** [comment]
One more thought — why not implement this as a wrapper around the existing torchmetrics classification metrics? It might help keep things more modular and aligned with standard usage.

**Comment by aymuos15** [comment]
> One more thought — why not implement this as a wrapper around the existing torchmetrics classification metrics? It might help keep things more modular and aligned with standard usage.
> 
> cc: [@Borda](https://github.com/Borda)

Yup, that sounds like the right way to go for me too. 

In the PR i will do that. 

Thanks a lot.

**Comment by rittik9** [comment]
> adding mixture that would have the formatting API and eventually some private tool so a relevant class will be extended with this mixing

agreed , I think this is the best way to approach this...

**Comment by rittik9** [comment]
Hi, @aymuos15 apologies for late reply. Your implementation also looks good to me.... pls feel free to open a pr

but it may raise some potential issues like 

1. Nested Metrics Issue:

```python 
self.precision = MulticlassPrecision(...)
self.recall = MulticlassRecall(...)
self.f1 = F1Score(...)
self.accuracy = Accuracy(...)
```

as  per tm docs:

> Due to specialized logic around metric states, we in general do not recommend that metrics are initialized inside other metrics (nested metrics), as this can lead to weird behaviour. Instead consider subclassing a metric or use torchmetrics.MetricCollection.

2. State management complexity:
The implementation maintains its own states (self.preds, self.target) alongside nested metric states, it creates multiple sources of truth and potential synchronization issues in distributed settings.The docs recommends using MetricCollection for managing multiple metrics.

3. Compute Inefficiency:
It doesn't leverage tm's `compute_groups` optimization.Recalculates the same statistics multiple times.

as per tm docs
```python
The compute_groups argument allow you to specify which metrics should share metric state. By default, this will automatically be derived but can also be set manually.

>>>
>>> metrics = MetricCollection(
...     MulticlassRecall(num_classes=3, average='macro'),
...     MulticlassPrecision(num_classes=3, average='macro'),
...     MeanSquaredError(),
...     compute_groups=[['MulticlassRecall', 'MulticlassPrecision'], ['MeanSquaredError']]
... )
>>> metrics.update(preds, target)
>>> pprint(metrics.compute())
{'MeanSquaredError': tensor(2.3750), 'MulticlassPrecision': tensor(0.0667), 'MulticlassRecall': tensor(0.1111)}
>>> pprint(metrics.compute_groups)
{0: ['MulticlassRecall', 'MulticlassPrecision'], 1: ['MeanSquaredError']}
```

keeping the above points in mind maybe you can modify your implementation like below:
```python
from torchmetrics import MetricCollection
from torchmetrics.classification import (
    MulticlassAccuracy, 
    MulticlassPrecision,
    MulticlassRecall,
    MulticlassF1Score
)

class ClassificationReport(Metric):
    def __init__(
        self,
        num_classes: int,
        **kwargs: Any,
    ) -> None:
        super().__init__(**kwargs)
        
        # Use MetricCollection for efficient computation
        self.metrics = MetricCollection({
            'precision': MulticlassPrecision(num_classes=num_classes, average=None),
            'recall': MulticlassRecall(num_classes=num_classes, average=None),
            'f1': MulticlassF1Score(num_classes=num_classes, average=None),
            'accuracy': MulticlassAccuracy(num_classes=num_classes, average='micro')
        }, compute_groups=[['precision', 'recall', 'f1']])  # Optimize computation
        
        self.num_classes = num_classes
        self.add_state("target", default=[], dist_reduce_fx="cat")

    def update(self, preds: Tensor, target: Tensor) -> None:
        self.metrics.update(preds, target)
        self.target.append(target)

    def compute(self) -> Union[Dict[str, Any], str]:
        metrics_dict = self.metrics.compute()
        support = torch.bincount(
            torch.cat(self.target), 
            minlength=self.num_classes
        ).float()
        
        # Format results...
```

**Comment by aymuos15** [comment]
Thank you very much for linking all the docs and the information. I will start a PR this week keeping these in mind.

**Comment by rittik9** [comment]
Hi @aymuos15 let me know once the pr is ready for review
